keps/sig-windows/windows-runtimeclass-support.md

```diff
@@ -0,0 +1,568 @@
+
+```yaml
+apiVersion: node.k8s.io/v1beta1
+kind: RuntimeClass
+metadata:
+  name: windows-1903
+handler: 'default'
+scheduling:
+  nodeSelector:
+    kubernetes.io/os: 'windows'
+    kubernetes.io/arch: 'amd64'
+    node.kubernetes.io/osversion: '10.0.18362'
+  tolerations:
+  - effect: NoSchedule
+    key: windows
+    operator: Equal
+    value: "true"
+```
+
+#### Story 3 - Choosing a specific multi-arch image
+
+Starting from story 2, the RuntimeClass `handler` field could also be used as a means os/version/architecture/variant used by the container runtime. The `handler` is matched up with a corresponding section in the ContainerD configuration file.
+
+Here's an example of what corresponding ContainerD configurations could look like for the runtimes above. The first one would force Hyper-V for compatibility with containers needing 10.0.17763. The second one would be the default for the current version where no Hyper-V isolation is required for compatibility.
+
+```toml
+        [plugins.cri.containerd.runtimes.containerd-hyperv-17763]
+          runtime_type = "io.containerd.runhcs.v1"
+          [plugins.cri.containerd.runtimes.containerd-hyperv-17763.options]
+            SandboxImage = "{{WINDOWSSANDBOXIMAGE}}"
+            SandboxPlatform = "windows/amd64"
+            SandboxOsVersion = "10.0.17763"
+            SandboxIsolation = 1
+
+        [plugins.cri.containerd.runtimes.default]
+          runtime_type = "io.containerd.runhcs.v1"
+          # No version is specified for process isolation, the node OS version is used
+```
+
+
+### Implementation Details/Notes/Constraints
+
+There were multiple options discussed with SIG-Node & SIG-Windows on October 8 2019 prior to filing this KEP. That discussion & feedback were captured in [Difficulties in mixed OS and arch clusters]. If you're looking for more details on other approaches excluded, please review that document.
+
+
+#### Adding new label node.kubernetes.io/osversion
+
+In [Bounding Self-Labeling Kubelets], a specific range of prefixes were reserved for node self-labeling - `[*.]node.kubernetes.io/*`. Adding a new label within that namespace won't require any changes to NodeRestriction admission. As a new field it also won't require changes to any existing workloads.
+
+Windows product names such as "Windows Server version 1903" don't align with the actual build numbers, so build numbers will be used instead. Users can choose to reflect the product name instead as shown in the user stories.
+
+[Windows Update History] has a full list of version numbers by release & patch. Starting from an example of `10.0.17763.805` - the OS major, minor, and build number - `10.0.17763` - should match for containers to be compatible. The final `.805` refers to the monthly patches and isn't required for compatibility. Therefore, a value such as `node.kubernetes.io/osversion = 10.0.17763` will be used.
```

Comment:
windows-build is fine w/ me.

Reply:
alright will change to windows-build. Thanks!

Reply:
does this version change on every windows update?

something similar was discussed for linux kernel version in https://github.com/kubernetes/kubernetes/pull/51006#issuecomment-327251049

were the need for the node to update the label value (as opposed to the set-once labels it sets now), and the issues with fine-grained selection discussed?

Reply:
> Therefore, a value such as `node.kubernetes.io/osversion = 10.0.17763` will be used.

Are we confident the version portion will always fit within label values validation?

> The final `.805` refers to the monthly patches and isn't required for compatibility. Therefore, a value such as `node.kubernetes.io/osversion = 10.0.17763` will be used.

Would a container ever have to target a specific fix level?

Reply:
(since the kernel version label discussion occurred, we're in a much better place with well-defined label prefixes nodes/kubelets can control, so the questions above are the only ones I have… I'll defer to node/windows reviewers on the suitability of labels for this purpose)

Reply:
great questions. Right now there's no use case where we'd have to target a specific patch level, but I suppose if someone wanted to do blue/green rollouts of host updates this could make sense.

Do you think adding a second one `node.kubernetes.io/windows-patch` makes sense? RuntimeClass should only generally constrain against `node.kubernetes.io/windows-build`

Reply:
Can you clarify what you mean by the labels values validation?

I could foresee a 11.0.xxxxx in the future. But since this is an exact string match, even if it were changed to 11.foo.0.1 that would still work as a label right?

Reply:
> Can you clarify what you mean by the labels values validation?
> 
> I could foresee a 11.0.xxxxx in the future. But since this is an exact string match, even if it were changed to 11.foo.0.1 that would still work as a label right?

As long as it still passed validation, yes, but label values allow a limited character set. Label values are validated with the following checks:

https://github.com/kubernetes/kubernetes/blob/release-1.16/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L30-L32:
```go
const qnameCharFmt string = "[A-Za-z0-9]"
const qnameExtCharFmt string = "[-A-Za-z0-9_.]"
const qualifiedNameFmt string = "(" + qnameCharFmt + qnameExtCharFmt + "*)?" + qnameCharFmt
```

https://github.com/kubernetes/kubernetes/blob/release-1.16/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L88-L108:
```go
const labelValueFmt string = "(" + qualifiedNameFmt + ")?"
const labelValueErrMsg string = "a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character"

// LabelValueMaxLength is a label's max length
const LabelValueMaxLength int = 63

var labelValueRegexp = regexp.MustCompile("^" + labelValueFmt + "$")

// IsValidLabelValue tests whether the value passed is a valid label value.  If
// the value is not valid, a list of error strings is returned.  Otherwise an
// empty list (or nil) is returned.
func IsValidLabelValue(value string) []string {
	var errs []string
	if len(value) > LabelValueMaxLength {
		errs = append(errs, MaxLenError(LabelValueMaxLength))
	}
	if !labelValueRegexp.MatchString(value) {
		errs = append(errs, RegexError(labelValueErrMsg, labelValueFmt, "MyValue", "my_value", "12345"))
	}
	return errs
}
```

If we're storing windows build versions directly in label values, we need to be sure that the build versions will not change in the future in ways that will no longer fit within the label value restrictions (length, other characters like `+`, etc), or if they do change, we have a plan to deal with translating those into allowed label values.

Reply:
yeah I think that validation should be ok for a few reasons:

1. The internal Windows build name (the full string) is limited to that character set. Anything we would use is a subset of that. Here's an example from Windows 10, but server follows the same format: `18362.1.amd64fre.19h1_release.190318-1202`
2. That version is not used as-is today. It's already massaged into 
a more readable format returned from `kubectl describe node` 
https://github.com/kubernetes/kubernetes/blob/0599ca2bcfcae7d702f95284f3c2e2c2978c7772/vendor/github.com/docker/docker/pkg/parsers/operatingsystem/operatingsystem_windows.go#L10 . For the kubelet's purpose, I'll be implementing something like that to not include the patch (UBR) number and instead just return the build number such as `10.0.17763`

Reply:
sounds reasonable. might be worth adding a note with those findings.

Reply:
sure, working on it now.

Reply:
added @ line 290 in latest revision

Reply:
added at line 290 below